model/histogram/float_histogram.go

```diff
@@ -647,80 +884,111 @@ func (b FloatBucket) String() string {
 
-	currCount            float64 // Count in the current bucket.
-	currIdx              int32   // The actual bucket index.
-	currLower, currUpper float64 // Limits of the current bucket.
-
-}
+	currCount float64 // Count in the current bucket.
+	currIdx   int32   // The bucket index within the targetSchema.
+	origIdx   int32   // The bucket index within the original schema.
 
-func newFloatBucketIterator(h *FloatHistogram, positive bool) *floatBucketIterator {
-	r := &floatBucketIterator{schema: h.Schema, positive: positive}
-	if positive {
-		r.spans = h.PositiveSpans
-		r.buckets = h.PositiveBuckets
-	} else {
-		r.spans = h.NegativeSpans
-		r.buckets = h.NegativeBuckets
-	}
-	return r
+	absoluteStartValue float64 // Never return buckets with an upper bound ≤ this value.
 }
 
-func (r *floatBucketIterator) Next() bool {
-	if r.spansIdx >= len(r.spans) {
+func (i *floatBucketIterator) Next() bool {
+	if i.spansIdx >= len(i.spans) {
 		return false
 	}
-	span := r.spans[r.spansIdx]
-	// Seed currIdx for the first bucket.
-	if r.bucketsIdx == 0 {
-		r.currIdx = span.Offset
-	} else {
-		r.currIdx++
-	}
-	for r.idxInSpan >= span.Length {
-		// We have exhausted the current span and have to find a new
-		// one. We'll even handle pathologic spans of length 0.
-		r.idxInSpan = 0
-		r.spansIdx++
-		if r.spansIdx >= len(r.spans) {
-			return false
+
+	// Copy all of these into local variables so that we can forward to the
+	// next bucket and then roll back if needed.
+	origIdx, spansIdx, idxInSpan := i.origIdx, i.spansIdx, i.idxInSpan
+	span := i.spans[spansIdx]
+	firstPass := true
+	i.currCount = 0
+
+mergeLoop: // Merge together all buckets from the original schema that fall into one bucket in the targetSchema.
+	for {
+		if i.bucketsIdx == 0 {
+			// Seed origIdx for the first bucket.
+			origIdx = span.Offset
+		} else {
+			origIdx++
+		}
+		for idxInSpan >= span.Length {
+			// We have exhausted the current span and have to find a new
+			// one. We even handle pathologic spans of length 0 here.
+			idxInSpan = 0
+			spansIdx++
+			if spansIdx >= len(i.spans) {
+				if firstPass {
+					return false
+				}
+				break mergeLoop
+			}
+			span = i.spans[spansIdx]
+			origIdx += span.Offset
+		}
+		currIdx := i.targetIdx(origIdx)
+		if firstPass {
+			i.currIdx = currIdx
+			firstPass = false
+		} else if currIdx != i.currIdx {
+			// Reached next bucket in targetSchema.
+			// Do not actually forward to the next bucket, but break out.
+			break mergeLoop
+		}
+		i.currCount += i.buckets[i.bucketsIdx]
+		idxInSpan++
+		i.bucketsIdx++
+		i.origIdx, i.spansIdx, i.idxInSpan = origIdx, spansIdx, idxInSpan
+		if i.schema == i.targetSchema {
+			// Don't need to test the next bucket for mergeability
+			// if we have no schema change anyway.
+			break mergeLoop
 		}
-		span = r.spans[r.spansIdx]
-		r.currIdx += span.Offset
 	}
+	// Skip buckets before absoluteStartValue.
+	// TODO(beorn7): Maybe do something more efficient than this recursive call.
```

Comment:
During the first pass of Next(), we can zoom through the initial buckets whose target idxs' bounds are below absoluteStartValue.

Reply:
That's what I tried initially. But there were weird corner cases that I found hard to resolve. The recursive appeared to me the logically cleanest solution. Perhaps let's try to tweak this in a separate PR (that's why I left the TODO).